src/main/archetype/ui.frontend/src/main/webpack/site/components/_helloworld.js

```diff
@@ -0,0 +1,88 @@
+
+// Example of how a component should be initialized via JavaScript
+// This script logs the value of the component's text property model message to the console
+
+(function() {
+    "use strict";
+
+    // Best practice:
+    // For a good separation of concerns, don't rely on the DOM structure or CSS selectors,
+    // but use dedicated data attributes to itentify all elements that the script needs to
+    // interacts with.
+    var selectors = {
+        self:      '[data-cmp-is="helloworld"]',
+        property:  '[data-cmp-hook-helloworld="property"]',
+        message:   '[data-cmp-hook-helloworld="model"]'
+    };
+
+    function HelloWorld(config) {
+        var that = this;
+
+        if (config && config.element) {
+            init(config);
+        }
+
+        function init(config) {
+            // Best practice:
+            // To prevents multiple initialization, remove the main data attribute that
+            // identified the component.
+            config.element.removeAttribute("data-cmp-is");
+
+            var property = config.element.querySelectorAll(selectors.property);
+            property = property.length == 1 ? property[0].textContent : null;
+
+            var model = config.element.querySelectorAll(selectors.message);
+            model = model.length == 1 ? model[0].textContent : null;
+
+            if (console && console.log) {
+                console.log(
+                    "HelloWorld component JavaScript example",
+                    "\nText property:\n", property,
+                    "\nModel message:\n", model
+                );
+            }
+        }
+    }
+
+    // Best practice:
+    // Use a method like this mutation obeserver to also properly initialize the component
+    // when an author drops it onto the page or modified it with the dialog.
```

Comment:
"mutation observer"

Just as a note here, while it's good to use the mutation observer, we should really only register one that handles all components. This is a known issue in the core components https://github.com/adobe/aem-core-wcm-components/issues/458

Reply:
@richardhand, indeed, once there is a proper js api we should remove the observation. Can you please create a new issue for this? Thanks!

Reply:
Also, this is a lot of boilerplate code to just initialize a component. We maybe should have a helper function or object that does that once and that every component can just use, whether it is a Core Component or a custom one.

Reply:
Totally agree, we sorely need a base component class for the UI widgets.